### PR TITLE
Added ESO in secrets

### PR DESCRIPTION
**Pull request by milindk8**
# Description

Add ESO (External Secret Operator) as an additional option while creating a secret.

Fixes - https://github.com/devtron-labs/devtron/issues/2169

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/secrets/Secret.tsx [hunk 1]
```diff
@@ -34,31 +36,227 @@ import { ReactComponent as Trash } from '../../assets/icons/ic-delete.svg'
...
+import { ReactComponent as InfoIcon } from '../../assets/icons/ic-info-i-outline.svg'
+import { getCustomOptionSelectionStyle } from '../v2/common/ReactSelect.utils'
 
-let sampleJSON = [
-    {
-        key: 'service/credentials',
-        name: 'secret-key',
-        property: 'property-name',
-        isBinary: true,
+export const sampleJSONs = {
```
Comment (vivek-devtron) [inline src/components/secrets/Secret.tsx:42]: Can we extract this sample JSON out to some separate file?

File: src/components/secrets/Secret.tsx [hunk 2]
```diff
@@ -34,31 +36,227 @@ import { ReactComponent as Trash } from '../../assets/icons/ic-delete.svg'
...
+        {
+            key: 'service/credentials',
+            name: 'secret-key',
+            property: 'property-name',
+            isBinary: true,
+        },
+    ],
+}
+
+export const dataHeaders = {
```
Comment (vivek-devtron) [inline src/components/secrets/Secret.tsx:171]: Same for this

File: src/components/secrets/Secret.tsx [hunk 3]
```diff
@@ -34,31 +36,227 @@ import { ReactComponent as Trash } from '../../assets/icons/ic-delete.svg'
...
+        <div>
+            # Sample Data<br></br># key: Secret key in backend<br></br># name: Name for this key in the generated secret
+            <br></br># property: Property to extract if secret in backend is a JSON object(optional)<br></br># isBinary: Set
+            this to true if configuring an item for a binary file stored(optional)<br></br>
+        </div>
+    )
+}
+
+
+export const getTypeGroups = (typeValue?: string) => {
```
Comment (vivek-devtron) [inline src/components/secrets/Secret.tsx:225]: Same heer

File: src/components/secrets/Secret.tsx [hunk 4]
```diff
@@ -256,6 +456,12 @@ export function ListComponent({ icon = '', title, subtitle = '', onClick, classN
         </article>
     )
 }
+
+interface EsoData {
```
Comment (vivek-devtron) [inline src/components/secrets/Secret.tsx:460]: extract out interfaces to some types file

File: src/components/secrets/Secret.tsx [hunk 5]
```diff
@@ -696,22 +972,55 @@ export const SecretForm: React.FC<SecretFormProps> = function (props) {
...
-                                {name}
-                            </Select.Option>
-                        ))}
-                    </Select>
+                    <ReactSelect 
+                    placeholder="Select Secret Type"
+                    options={getTypeGroups()}
+                    defaultValue={externalType && externalType !== '' ? getTypeGroups(externalType) : getTypeGroups()[0].options[0]}
+                    onChange={onChange}
+                    styles={{
```
Comment (vivek-devtron) [inline src/components/secrets/Secret.tsx:980]: extract out this style

File: src/components/secrets/Secret.tsx [hunk 6]
```diff
@@ -696,22 +972,55 @@ export const SecretForm: React.FC<SecretFormProps> = function (props) {
...
+                            marginBottom: 0,
+                        }),
+                        indicatorsContainer: (provided, state) => ({
+                            ...provided,
+                        })
+                    }}
+                    components={{
+                        IndicatorSeparator: null,
+                        Option: (props) => {
+                            props.selectProps.styles.option = getCustomOptionSelectionStyle()
```
Comment (vivek-devtron) [inline src/components/secrets/Secret.tsx:1005]: same for the components. please extract out and create separate function.

File: src/components/secrets/Secret.tsx (hunk 1 above)
Comment (milindk8) [reply src/components/secrets/Secret.tsx]: Done

File: src/components/secrets/Secret.tsx (hunk 3 above)
Comment (milindk8) [reply src/components/secrets/Secret.tsx]: Done

File: src/components/secrets/Secret.tsx (hunk 4 above)
Comment (milindk8) [reply src/components/secrets/Secret.tsx]: Done

File: src/components/secrets/Secret.tsx (hunk 6 above)
Comment (milindk8) [reply src/components/secrets/Secret.tsx]: Created

File: src/components/EnvironmentOverride/SecretOverrides.tsx [hunk 7]
```diff
@@ -269,14 +246,26 @@ export function OverrideSecretForm({ name, appChartRef, toggleCollapse }) {
         PATTERNS.CONFIG_MAP_AND_SECRET_KEY,
         `Key must consist of alphanumeric characters, '.', '-' and '_'`,
     )
+    const isEsoSecretData = esoSecretData?.secretStore && esoSecretData?.esoData.length > 0
     const [yamlMode, toggleYamlMode] = useState(true)
+    const [esoDataSecret, setEsoData] = useState(esoSecretData?.esoData)
+    const [secretStore, setSecretStore] = useState(esoSecretData?.secretStore)
     const [isFilePermissionChecked, setIsFilePermissionChecked] = useState(!!filePermission)
+    const [esoSecretYaml, setEsoYaml] = useState(YAML.stringify(isEsoSecretData ? esoSecretData : {}))
+    const sample = YAML.stringify(sampleJSONs[externalType] || sampleJSONs["default"])
```
Comment (vivek-devtron) [inline src/components/EnvironmentOverride/SecretOverrides.tsx:255]: Please use ' in tsx i.e. "default" to 'default'

File: src/components/EnvironmentOverride/SecretOverrides.tsx [hunk 8]
```diff
@@ -546,7 +555,25 @@ export function OverrideSecretForm({ name, appChartRef, toggleCollapse }) {
...
+                    }
+                    if (j.property) {
+                        temp['property'] = j.property
+                    }
+                    return temp
+                })
+                setEsoData(jsonList)
+            }
+        } catch (error) {
+            // console.log(error)
```
Comment (vivek-devtron) [inline src/components/EnvironmentOverride/SecretOverrides.tsx:575]: please remove console.log

**Review by vivek-devtron — APPROVED** [review]

File: src/components/EnvironmentOverride/SecretOverrides.tsx (hunk 8 above)
Comment (milindk8) [reply src/components/EnvironmentOverride/SecretOverrides.tsx]: done

File: src/components/EnvironmentOverride/SecretOverrides.tsx (hunk 7 above)
Comment (milindk8) [reply src/components/EnvironmentOverride/SecretOverrides.tsx]: Done

File: src/components/EnvironmentOverride/SecretOverrides.tsx [hunk 9]
```diff
@@ -269,14 +246,26 @@ export function OverrideSecretForm({ name, appChartRef, toggleCollapse }) {
         PATTERNS.CONFIG_MAP_AND_SECRET_KEY,
         `Key must consist of alphanumeric characters, '.', '-' and '_'`,
     )
+    const isEsoSecretData = esoSecretData?.secretStore && esoSecretData?.esoData.length > 0
```
Comment (sohel-rp) [inline src/components/EnvironmentOverride/SecretOverrides.tsx:249]: Just first time option check will do or else do something like  `esoSecretData && esoSecretData.secretStore ....`

Comment (milindk8) [reply src/components/EnvironmentOverride/SecretOverrides.tsx]: done

File: src/components/EnvironmentOverride/SecretOverrides.tsx [hunk 10]
```diff
@@ -522,14 +529,16 @@ export function OverrideSecretForm({ name, appChartRef, toggleCollapse }) {
 
     function handleSecretDataYamlChange(yaml): void {
         if (codeEditorRadio !== 'data') return
-        setSecretDataYaml(yaml)
+        isESO ? setEsoYaml(yaml) : setSecretDataYaml(yaml)
```
Comment (sohel-rp) [inline src/components/EnvironmentOverride/SecretOverrides.tsx:532]: Use if...else

File: src/components/EnvironmentOverride/SecretOverrides.tsx [hunk 11]
```diff
@@ -546,7 +555,24 @@ export function OverrideSecretForm({ name, appChartRef, toggleCollapse }) {
                 })
                 setSecretData(json)
             }
-        } catch (error) {}
+            if(isESO && Array.isArray(json?.esoData)){
+                let jsonList = json?.esoData.map((j) => {
```
Comment (sohel-rp) [inline src/components/EnvironmentOverride/SecretOverrides.tsx:559]: Use const

File: src/components/EnvironmentOverride/SecretOverrides.tsx [hunk 12]
```diff
@@ -546,7 +555,24 @@ export function OverrideSecretForm({ name, appChartRef, toggleCollapse }) {
...
+                    let temp = {}
+                    if (j.secretKey) {
```
Comment (sohel-rp) [inline src/components/EnvironmentOverride/SecretOverrides.tsx:561]: can't we simply return `j` or are there any additional properties that we don't want?

File: src/components/deploymentConfig/types.tsx [hunk 13]
```diff
@@ -71,3 +71,30 @@ export interface DeploymentConfigProps extends RouteComponentProps<DeploymentCon
...
+export interface SecretFormProps {
+    id: number
+    appChartRef: { id: number; version: string; name: string }
+    appId: number
+    roleARN: string
+    name: string
+    index: number
+    external: boolean
+    externalType: string
+    secretData: { key: string; name: string; property: string; isBinary: boolean }[]
```
Comment (sohel-rp) [inline src/components/deploymentConfig/types.tsx:88]: Create separate type for this

File: src/components/deploymentConfig/types.tsx [hunk 14]
```diff
@@ -71,3 +71,30 @@ export interface DeploymentConfigProps extends RouteComponentProps<DeploymentCon
...
+    esoSecretData?: { secretStore: any; esoData: EsoData[] }
```
Comment (sohel-rp) [inline src/components/deploymentConfig/types.tsx:89]: same here

File: src/components/secrets/Secret.tsx [hunk 15]
```diff
@@ -629,15 +630,18 @@ export const SecretForm: React.FC<SecretFormProps> = function (props) {
     }
 
     function handleSecretDataYamlChange(yaml): void {
+        
         if (codeEditorRadio !== 'data') return
-        setSecretDataYaml(yaml)
+        isESO ? setEsoYaml(yaml) : setSecretDataYaml(yaml)
```
Comment (sohel-rp) [inline src/components/secrets/Secret.tsx:635]: same here

File: src/components/secrets/Secret.tsx [hunk 16]
```diff
@@ -654,6 +658,22 @@ export const SecretForm: React.FC<SecretFormProps> = function (props) {
                 })
                 setSecretData(json)
             }
+            if(isESO && Array.isArray(json?.esoData)){
+                let jsonList = json?.esoData.map((j) => {
```
Comment (sohel-rp) [inline src/components/secrets/Secret.tsx:662]: same here

Comment (sohel-rp) [reply src/components/secrets/Secret.tsx:662]: If possible move to separate function and use at both places

Comment (milindk8) [reply src/components/secrets/Secret.tsx]: Next task this function will be different so not creating common function for this

File: src/components/secrets/Secret.tsx (hunk 15 above)
Comment (milindk8) [reply src/components/secrets/Secret.tsx]: done

File: src/components/deploymentConfig/types.tsx (hunk 14 above)
Comment (milindk8) [reply src/components/deploymentConfig/types.tsx]: done

File: src/components/deploymentConfig/types.tsx (hunk 13 above)
Comment (milindk8) [reply src/components/deploymentConfig/types.tsx]: Done

File: src/components/EnvironmentOverride/SecretOverrides.tsx (hunk 11 above)
Comment (milindk8) [reply src/components/EnvironmentOverride/SecretOverrides.tsx]: Done

File: src/components/EnvironmentOverride/SecretOverrides.tsx (hunk 10 above)
Comment (milindk8) [reply src/components/EnvironmentOverride/SecretOverrides.tsx]: Done

File: src/components/EnvironmentOverride/SecretOverrides.tsx (hunk 12 above)
Comment (milindk8) [reply src/components/EnvironmentOverride/SecretOverrides.tsx:565]: New task will change all the working of this function

Comment (milindk8) [reply src/components/EnvironmentOverride/SecretOverrides.tsx:565]: Not making any changes for now

**Review by sohel-rp — APPROVED** [review]

**Pull request merged**